---
title: "CellAssign_final"
author: "Shaila Fye"
date: "6/9/2020"
output: html_document
---
- need to revise puram data set to remove the first 5 rows
- re-write this to another file
- use this file and create a sce object
- then do cell assign feature 

https://kieranrcampbell.github.io/r-workshop-march-2019/#using_cellassign_to_assign_cells_to_known_types


Save annotations
```{r Read in data and parse}
data <- read.delim(file ="/Users/shailafye/Documents/summer2020/puram/GSE103322_HNSCC_all_data (1).txt", comment = "#", stringsAsFactors=F)

head(data)
#save annotaions (first 5 rows)
anno <- data[1:5,]
anno
write.table(anno,file="/Users/shailafye/Documents/summer2020/puram/annotaions_head_and_neck_all_data.txt", sep = "\t")
#remove first 5 rows
data <- tail(data,-5)
head(data)

#remove quotation marks and make the rownames the name of the first column and then remove the first column
row.names(data) = gsub('\'','',data[,1])
data_parsed = data[,-1]



#data_parsed is final data set

```
data_parsed is final data set
This is with the annotations removed
and this is with the rows set to the gene name
now we are going to write this to a new file
```{r}
data_parsed<- data.frame(data_parsed)
write.table(data_parsed,file="/Users/shailafye/Documents/summer2020/puram/head_and_neck_all_data.txt", sep = "\t")
```

Now we want to open this new data set (its been parsed appropriately)
Create a sce object and run cellAssign

```{r read parsed data file in }
myData <- read.delim(file="/Users/shailafye/Documents/summer2020/puram/head_and_neck_all_data.txt", comment = "#", stringsAsFactors=F)

str(myData)
head(myData)

```

Create a SCE object
```{r sce object}
library(SingleCellExperiment)

sce <- SingleCellExperiment(list(counts=as.matrix(myData)))
head(sce)
cell.labels <- colnames(myData)

sc <- SingleCellExperiment(
    list(counts=as.matrix(myData)),
    colData=DataFrame(label=cell.labels),
    rowData = DataFrame(length=rownames(myData)),
    metadata=list(study="GSE103322")
)

head(sc)
colData(sc)
rownames(sc)
rowData(sc)

rownames(sc)


```

3. Prepare single-cell expression data in the form of a SingleCellExperiment object
https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html
We will assume this object is “sce”. In rowData(sce) should be fields “ID”, corresponding to ensembl
gene ID, and “Symbol”, corresponding to HGNC symbol.

 Compute size factors using scran
 
```{r sumfactor}

library(scran)

sce <- computeSumFactors(sce)


```


#This is in the form of a list, where the names of the list are the cell types and the contents are marker genes for the cell types. An example can be found in the CellAssign package by calling
```{r import tme markers}
library(cellassign)
data(example_TME_markers)
str(example_TME_markers)
length(example_TME_markers)


#Turn marker list into binary matrix using marker_list_to_mat

#marker_mat <- marker_list_to_mat(example_TME_markers, include_other = TRUE)
marker_mat <- marker_list_to_mat(example_TME_markers$symbol)

marker_mat
length(marker_mat)

#Optional: an “unknown” cell type may be included by passing include_other = TRUE to
#marker_list_to_mat


```

Attempt to match marker genes to data sce object
```{r}
#Match IDs to rows of the SingleCellExperiment
matched <- match(rownames(marker_mat), rowData(sc)$length)
class(matched)
matched
matched[is.na(matched)] <- 0

#genes<-myData[,1]
rowData(sc)$length

rownames(marker_mat)
colnames(marker_mat)
marker_mat
#Subset SingleCellExperiment to markers only

sce_marker <- sce[matched,]
#sce_marker <- sce[intersect(rownames(marker_mat), rownames(sce)),]
#Note that the rows in the single cell experiment or gene expression matrix should be ordered identically to those in the marker input matrix.



```

Run Cell Assign
Check if tensorflow is imported correctly

```{r}
sce_marker
install.packages("tensorflow", repos = "http://cran.us.r-project.org")

tensorflow::install_tensorflow()
library(tensorflow)


install.packages("tensorflow")
library(tensorflow)
install_tensorflow(extra_packages = "tensorflow-probability")


fit <- cellassign( exprs_obj = sce_marker, 
                   marker_gene_info = marker_mat, 
                   s = sizeFactors(sce_marker))
x<- fit$cell_type
z <- fit$mle_params
print(fit$cell_type)
print(fit$mle_params)
#maximum likelihood estimation (MLE) 

fit

class(z)
v<-as.data.frame(z)
v<-data.matrix(z)
v
head(v)
str(v)
class(v)

#save this data

write.table(x,file="/Users/shailafye/Documents/summer2020/puram/cell_type", sep = "\t")
#write(z,file="/Users/shailafye/Documents/summer2020/puram/mle_params.txt")

lapply(z, write, "/Users/shailafye/Documents/summer2020/puram/mle_params.txt", append=TRUE)
#?number of columns


```

Object fit:
A cellassign fit for 5902 cells, 39 genes, 9 cell types with 0 covariates
            To access cell types, call celltypes(x)
            To access cell type probabilities, call cellprobs(x)
```{r}
prob <- cellprobs(fit)
class(prob)
write.table(prob,file="/Users/shailafye/Documents/summer2020/puram/cell_type_probability.txt", sep = "\t")
type <- celltypes(fit)


```



*Extra:
checking annotations of published data
- changing the 0 to cancer or other...
- then we can put it on the tsne plot to compare with the data we generated through cell assign
"/Users/shailafye/Documents/summer2020/puram/annotaions_head_and_neck_all_data.txt"
```{r}
anno<- read.delim(file="/Users/shailafye/Documents/summer2020/puram/annotaions_head_and_neck_all_data.txt", stringsAsFactors=F)
head(anno)

ct <- anno[5,]

head (ct)

# for loop?
# if(ct == 0 && anno[3,] == 1), replace that zero with "cancer cell" --> if anno[3,] == 0, then replace with "other"

colnames(ct) <- NULL
ct
ct1 <- ct[,-1]
row.names(ct1) = ct1[1]
#ct1 <- ct1[,-0]
class(ct1)
ct_published <- as.character(ct1[1,])

class(ct_published) 

head(ct_published)
str(ct_published)
table(ct_published)


cell_type_dataframe <- as.data.frame(ct_published)
head(cell_type_dataframe)

#now replace the 0 with cancer cell if row 3 in anno is a 1

c <- anno[3,]

colnames(c) <- NULL
c
c1 <- c[,-1]
row.names(c1) = c1[1]
class(c1)
cancer_anno <- as.character(c1[1,])
class(cancer_anno) 
head(cancer_anno)
str(cancer_anno)

cancer_dataframe <- as.data.frame(cancer_anno)
head(cancer_dataframe)

#replace 0 with "cancer cell" if there is a 1 in cancer_dataframe

write.table(cell_type_dataframe,file="/Users/shailafye/Documents/summer2020/puram/cell_type_dataframe", sep = "\t")
write.table(cancer_dataframe,file="/Users/shailafye/Documents/summer2020/puram/cancer_dataframe", sep = "\t")

#change the other to cancer cell if it was detected
modified_list <- c()
i <- 1
for(val in ct_published ){
    if (val == "0"){
        if(cancer_anno[i] == "1"){
            modified_list[i] <- c("Cancer Cell")
        }else if(cancer_anno[i] == "0"){
            modified_list[i] <- c("Other")
        }
    } else {
        modified_list[i] <- val
    }
    i <- i+1
}


class(modified_list) #character vector

which(is.na(modified_list))
cancer_anno[62]
ct_published[62]
table(modified_list)

as.data.frame(modified_list)
write.table(modified_list,file="/Users/shailafye/Documents/summer2020/puram/published_celltype", sep = "\t")


```


Combining lists w/ cell barcode and both published and cell assign data to compare in UpSet
```{r}
published <- read.delim(file ="/Users/shailafye/Documents/summer2020/puram/published_celltype")
cell_assign <- read.delim(file ="/Users/shailafye/Documents/summer2020/puram/cell_type")
names <- colnames(data)
names[1]<- "Cell Barcodes"
head(names)
names <- as.data.frame(names)
colnames(names) <- "Cell Barcodes"	
names<- names[-1,]
class(names)


published[1]
rownames(published) <- NULL
colnames(published) <- "Cell Type Published"
as.vector(published)
class(published)

cell_assign[1]
rownames(cell_assign) <- NULL
colnames(cell_assign) <- "Cell Assign"
as.vector(cell_assign)
#cell_assign<-class(cell_assign)



cell_type_comparisons <- data.frame("Cell Barcode" = names,"Cell Type Published" =published , "Cell Assign" = cell_assign )

write.table(cell_type_comparisons,file="/Users/shailafye/Documents/summer2020/puram/cell_type_comparisons", sep = "\t")

```

