---
title: "CellAssign_regev"
author: "Shaila Fye"
date: "7/30/2020"
output: html_document
---

Source: https://kieranrcampbell.github.io/r-workshop-march-2019/#using_cellassign_to_assign_cells_to_known_types

write in file paths to run program
files:
```{r}
#count matrix (normalized and filtered prior)
allData <- "/Users/shailafye/Documents/summer2020/regev/regev_alldata.txt" #insert path to txt file with count matrix (genes (row) x cell (column))

#extract annotations, path to save information in a new file
annotationData <- " " #NA
#path to crate a new txt file with extracted annotations
#EX: /Users/shailafye/Documents/summer2020/puram/annotaions_head_and_neck_all_data.txt

#file name to save parsed data file (without annotations and corrected rownames)
#creating this file!
finalData <- allData

#create a new file to store cell type data from cell assignment
cellTypeData <- "/Users/shailafye/Documents/summer2020/regev/CellAssign_regev"
# ex: "/Users/shailafye/Documents/summer2020/puram/cell_type"


#file to save cellAssign probability data
probData <- "/Users/shailafye/Documents/summer2020/regev/cellassign_cell_type_probability.txt"
#example: "/Users/shailafye/Documents/summer2020/puram/cell_type_probability.txt"

```



Read in Data
- if annotations are within data, remove and save separately
```{r Read in data and parse}
data <- read.delim(file = allData, comment = "#", stringsAsFactors=F)
head(data)
dim(data)
#23686 genes  6826 cells

```

Now we want to open this new data set (its been parsed appropriately)
Create a sce object and run cellAssign

```{r read parsed data file in }
#myData <- read.delim(file= finalData , comment = "#", stringsAsFactors=F)
myData <- data
str(myData)
head(myData)

```

Create a SCE object
```{r sce object}
library(SingleCellExperiment)

sce <- SingleCellExperiment(list(counts=as.matrix(myData)))
head(sce)
cell.labels <- colnames(myData)

sc <- SingleCellExperiment(
    list(counts=as.matrix(myData)),
    colData=DataFrame(label=cell.labels),
    rowData = DataFrame(length=rownames(myData)),
    metadata=list(study="GSE115978") #change study identifier!!
    )

head(sc)
colData(sc)
rownames(sc)
rowData(sc)

rownames(sc)



```

3. Prepare single-cell expression data in the form of a SingleCellExperiment object
https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html
We will assume this object is “sce”. In rowData(sce) should be fields “ID”, corresponding to ensembl
gene ID, and “Symbol”, corresponding to HGNC symbol.

 Compute size factors using scran
 
```{r sumfactor}

library(scran)

sce <- computeSumFactors(sce)


```


#This is in the form of a list, where the names of the list are the cell types and the contents are marker genes for the cell types. An example can be found in the CellAssign package by calling
```{r import tme markers}
library(cellassign)
data(example_TME_markers)
str(example_TME_markers)
length(example_TME_markers)


#Turn marker list into binary matrix using marker_list_to_mat

#marker_mat <- marker_list_to_mat(example_TME_markers, include_other = TRUE)
marker_mat <- marker_list_to_mat(example_TME_markers$symbol)

marker_mat
length(marker_mat)

#Optional: an “unknown” cell type may be included by passing include_other = TRUE to
#marker_list_to_mat


```

Attempt to match marker genes to data sce object
```{r}
#Match IDs to rows of the SingleCellExperiment
matched <- match(rownames(marker_mat), rowData(sc)$length)
class(matched)
matched
matched[is.na(matched)] <- 0

#genes<-myData[,1]
rowData(sc)$length

rownames(marker_mat)
colnames(marker_mat)
marker_mat
#Subset SingleCellExperiment to markers only

sce_marker <- sce[matched,]
#sce_marker <- sce[intersect(rownames(marker_mat), rownames(sce)),]
#Note that the rows in the single cell experiment or gene expression matrix should be ordered identically to those in the marker input matrix.



```

Run Cell Assign
Check if tensorflow is imported correctly

```{r}
sce_marker
library(tensorflow)

#do if not installed already
install.packages("tensorflow", repos = "http://cran.us.r-project.org")
tensorflow::install_tensorflow()
library(tensorflow)
install.packages("tensorflow")
library(tensorflow)
install_tensorflow(extra_packages = "tensorflow-probability")


fit <- cellassign( exprs_obj = sce_marker, 
                   marker_gene_info = marker_mat, 
                   s = sizeFactors(sce_marker))
x<- fit$cell_type
z <- fit$mle_params
print(fit$cell_type)
print(fit$mle_params)
#maximum likelihood estimation (MLE) 

table(x)
fit


#save this data
#can save other parameters/components like probability info from object fit
#"/Users/shailafye/Documents/summer2020/puram/cell_type" example file save
write.table(x,file=cellTypeData, sep = "\t")



```

Object fit:
A cellassign fit for 5902 cells, 39 genes, 9 cell types with 0 covariates
            To access cell types, call celltypes(x)
            To access cell type probabilities, call cellprobs(x)
```{r}
prob <- cellprobs(fit)
class(prob)
write.table(prob,file= probData, sep = "\t")
#"/Users/shailafye/Documents/summer2020/puram/cell_type_probability.txt" example file store
type <- celltypes(fit)


```


