---
title: "SingleR_regev"
author: "Shaila Fye"
date: "7/30/2020"
output: html_document
---

http://127.0.0.1:13952/library/SingleR/doc/SingleR.html
https://bioconductor.org/packages/release/bioc/html/SingleR.html
https://nbisweden.github.io/excelerate-scRNAseq/session-celltypeid/celltypeid.html


write in file paths to run program
files:
```{r}
#count matrix (normalized and filtered prior)
finalData <- "/Users/shailafye/Documents/summer2020/regev/regev_alldata.txt"#insert path to txt file with count matrix (genes (row) x cell (column))
#no annotaions in this dataset, row.names are the genes 


#create a new file to store cell type data from cell assignment
cellTypeData <- "/Users/shailafye/Documents/summer2020/regev/revev_celltype_singleR.txt" 
# ex: "/Users/shailafye/Documents/summer2020/puram/celltype_singleR_1"

```






install singleR
```{r}
library(SingleR)
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("SingleR")
library(SingleR)

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
#--------------

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("scRNAseq")

```

http://127.0.0.1:13952/library/SingleR/doc/SingleR.html

SingleR provides several reference datasets (mostly derived from bulk RNA-seq or microarray data) through dedicated data retrieval functions. For example, we obtain reference data from the Human Primary Cell Atlas using the HumanPrimaryCellAtlasData() function, which returns a SummarizedExperiment object containing matrix of log-expression values with sample-level labels.

https://version.helsinki.fi/hpessa/single-cell-rna-seq-data-analysis-2019/blob/7c0a5b301c8545166c6e2b936675122d6de04b24/Day2/celltypeid.Rmd (OTHER REFERNCE LISTS)
We will now do our cell type identification using SingleR. SingleR
comes with a collection of reference data sets. There are two for human:
`hpca`, the Human Primary Cell Atlas (microarray-based), and
`blueprint\_encode`, a combined Blueprint Epigenomics and Encode data
set (RNASeq based) .

For mouse there are also two sets: `immgen`, the Immunological Genome
Project (microarray-based) and `mouse.rnaseq`, a brain specific reference
(RNASeq-based). For details I refer to the SingleR-specifications
vignette.

Each reference data set comes in two 'flavours': 'types', which are very
fine-grained, detailed types, and 'main_types', which are less coarser
subset of those types.

The SingleR reference data sets are part of the package, and can be
explored easily.


```{r}
library(SingleR)



#this is the marker list... i can see if i can find another one
hpca.se <- HumanPrimaryCellAtlasData()
str(hpca.se)
head(hpca.se)

table(hpca.se$label.main)
table(hpca.se$label.fine)



data<- read.delim(file=finalData, sep = "\t")
#"/Users/shailafye/Documents/summer2020/puram/head_and_neck_all_data.txt"

```


Create a sce object

```{r}
library(SingleCellExperiment)

sce <- SingleCellExperiment(list(counts=as.matrix(data)))
head(sce)
cell.labels <- colnames(data)

sc <- SingleCellExperiment(
    list(counts=as.matrix(data)),
    colData=DataFrame(label=cell.labels),
    rowData = DataFrame(length=rownames(data)),
    metadata=list(study="GSE103322")
)

head(sc)
colData(sc)
rownames(sc)
rowData(sc)

library(scater)
scc <- logNormCounts(sc)

```
We use our hpca.se reference to annotate each cell in hESCs via the SingleR() function, which uses the algorithm described above. Note that the default marker detection method is to take the genes with the largest positive log-fold changes in the per-label medians for each gene.



```{r}
pred.hesc <- SingleR(test = scc, ref = hpca.se, labels = hpca.se$label.main)
pred.hesc
head(pred.hesc)
```

Each row of the output DataFrame contains prediction results for a single cell. Labels are shown before fine-tuning (first.labels), after fine-tuning (labels) and after pruning (pruned.labels), along with the associated scores. We summarize the distribution of labels across our subset of cells:

```{r}
head(pred.hesc)
colnames(pred.hesc)
head(pred.hesc$labels)
table(pred.hesc$labels) #why are there so many neurons
write.table(pred.hesc$labels,file= cellTypeData, sep = "\t")
#"/Users/shailafye/Documents/summer2020/puram/celltype_singleR_1"
```





