---
title: "tirosh_CellAssign"
author: "Shaila Fye"
date: "7/28/2020"
output: html_document
---
Create a sce object and run cellAssign
/Users/shailafye/Documents/summer2020/tirosh/melanoma_tirosh_alldata --> all data
/Users/shailafye/Documents/summer2020/tirosh/melanoma_published_celltypes --> published cell types
CELL ASSIGN DATA -->  "/Users/shailafye/Documents/summer2020/tirosh/cellAssign_tirosh_cell_type"

https://kieranrcampbell.github.io/r-workshop-march-2019/#using_cellassign_to_assign_cells_to_known_types
/Users/shailafye/Documents/summer2020/CellAssign_manuscript.pdf
- Assigning scRNA-seq data to known and de novo cell types using CellAssign

```{r read parsed data file in }
myData <- read.delim(file="/Users/shailafye/Documents/summer2020/tirosh/melanoma_tirosh_alldata", comment = "#", stringsAsFactors=F)

str(myData)
head(myData)

```

Create a SCE object
```{r sce object}
library(SingleCellExperiment)

sce <- SingleCellExperiment(list(counts=as.matrix(myData)))
head(sce)
cell.labels <- colnames(myData)

sc <- SingleCellExperiment(
    list(counts=as.matrix(myData)),
    colData=DataFrame(label=cell.labels),
    rowData = DataFrame(length=rownames(myData)),
    metadata=list(study="GSE72056")
)

head(sc)
colData(sc)
rownames(sc)
rowData(sc)

rownames(sc)


```

3. Prepare single-cell expression data in the form of a SingleCellExperiment object
https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html
We will assume this object is “sce”. In rowData(sce) should be fields “ID”, corresponding to ensembl
gene ID, and “Symbol”, corresponding to HGNC symbol.

 Compute size factors using scran
 
```{r sumfactor}

library(scran)

sce <- computeSumFactors(sce)


```


#This is in the form of a list, where the names of the list are the cell types and the contents are marker genes for the cell types. An example can be found in the CellAssign package by calling
```{r import tme markers}
library(cellassign)
data(example_TME_markers)
str(example_TME_markers)
length(example_TME_markers)


#Turn marker list into binary matrix using marker_list_to_mat

#marker_mat <- marker_list_to_mat(example_TME_markers, include_other = TRUE)
marker_mat <- marker_list_to_mat(example_TME_markers$symbol)

marker_mat
length(marker_mat)

#Optional: an “unknown” cell type may be included by passing include_other = TRUE to
#marker_list_to_mat


```

Attempt to match marker genes to data sce object
```{r}
#Match IDs to rows of the SingleCellExperiment
matched <- match(rownames(marker_mat), rowData(sc)$length)
class(matched)
matched
matched[is.na(matched)] <- 0

#genes<-myData[,1]
rowData(sc)$length 

rownames(marker_mat)
colnames(marker_mat)
marker_mat
#Subset SingleCellExperiment to markers only

sce_marker <- sce[matched,]
#sce_marker <- sce[intersect(rownames(marker_mat), rownames(sce)),]
#Note that the rows in the single cell experiment or gene expression matrix should be ordered identically to those in the marker input matrix.



```

Run Cell Assign
Check if tensorflow is imported correctly

```{r}
sce_marker
install.packages("tensorflow", repos = "http://cran.us.r-project.org")

tensorflow::install_tensorflow()
library(tensorflow)


install.packages("tensorflow")
library(tensorflow)
install_tensorflow(extra_packages = "tensorflow-probability")

#appx 10 min to run 
fit <- cellassign( exprs_obj = sce_marker, 
                   marker_gene_info = marker_mat, 
                   s = sizeFactors(sce_marker))
x<- fit$cell_type
z <- fit$mle_params
print(fit$cell_type)
print(fit$mle_params)
#maximum likelihood estimation (MLE) 
table(fit$cell_type)
fit

class(z)

head(x)
#save this data

write.table(x,file="/Users/shailafye/Documents/summer2020/tirosh/cellAssign_tirosh_cell_type", sep = "\t")
#write(z,file="/Users/shailafye/Documents/summer2020/puram/mle_params.txt")

lapply(z, write, "/Users/shailafye/Documents/summer2020/tirosh/cellAssign_tirosh_mle_params.txt", append=TRUE)
#?number of columns


```

Object fit:
A cellassign fit for 5902 cells, 39 genes, 9 cell types with 0 covariates
            To access cell types, call celltypes(x)
            To access cell type probabilities, call cellprobs(x)
```{r}
prob <- cellprobs(fit)
class(prob)
write.table(prob,file="/Users/shailafye/Documents/summer2020/tirosh/tirosh_cellAssign_cell_type_probability.txt", sep = "\t")
type <- celltypes(fit)



```


