---
title: "tsne_umap_tirosh_cellAssign"
author: "Shaila Fye"
date: "7/28/2020"
output: html_document
---
- tirosh data set being used -- /Users/shailafye/Documents/summer2020/tirosh/melanoma_tirosh_alldata
- cell assign

Based on seurat package in R

https://rpubs.com/mathetal/clusterlabel
https://www.youtube.com/watch?v=UbOlhhpNj-A&list=PLaZuFfQk5gXPuXNA00R_uEfD422lmkSnr&index=2 

umap: https://satijalab.org/seurat/v3.1/pbmc3k_tutorial.html 

https://scrnaseq-course.cog.sanger.ac.uk/website/seurat-chapter.html#run-non-linear-dimensional-reduction-tsne

```{r}
library(SingleCellExperiment)
#all data without annotations
r1 = read.table ("/Users/shailafye/Documents/summer2020/tirosh/melanoma_tirosh_alldata", header= TRUE, row.names = 1)
str(r1)
head(r1)
class(r1)
sce = SingleCellExperiment(list(counts = data.matrix(r1)))
library(scran)

clusters = quickCluster(sce, min.size=100)
sce = computeSumFactors (sce, cluster = clusters)

#sce = normalize (sce, return_log = FALSE)
#already normalized data

library(Seurat)

#create Seurat object using scran object
s_obj= CreateSeuratObject(counts = log(counts(sce) +1))
#find variable features to reduce run time
s_obj=FindVariableFeatures(s_obj)
#regress out batch
s_obj_nobatch = ScaleData(s_obj)
s_obj = RunPCA(s_obj_nobatch, features = VariableFeatures(s_obj))

s_obj = FindNeighbors(s_obj)
s_obj = FindClusters(s_obj, resolution = 0.15)

head(s_obj)
table(Idents(s_obj))

#tsne
s_obj = RunTSNE(s_obj)

DimPlot (s_obj, reduction = "tsne", label = TRUE)

saveRDS(s_obj, file = "/Users/shailafye/Documents/summer2020/tirosh/tirosh_tnse_attempt.rds")


# If you haven't installed UMAP, you can do so via reticulate::py_install(packages =
# 'umap-learn')

pbmc <- RunUMAP(s_obj, dims = 1:10)
# note that you can set `label = TRUE` or use the LabelClusters function to help label
# individual clusters
DimPlot(pbmc, reduction = "umap")
saveRDS(pbmc, file = "/Users/shailafye/Documents/summer2020/tirosh/tirosh_umap_attempt.rds")




```

Finding differentially expressed features (cluster biomarkers)
Seurat can help you find markers that define clusters via differential expression. By default, it identifes positive and negative markers of a single cluster (specified in ident.1), compared to all other cells. FindAllMarkers automates this process for all clusters, but you can also test groups of clusters vs. each other, or against all cells.

The min.pct argument requires a feature to be detected at a minimum percentage in either of the two groups of cells, and the thresh.test argument requires a feature to be differentially expressed (on average) by some amount between the two groups. You can set both of these to 0, but with a dramatic increase in time - since this will test a large number of features that are unlikely to be highly discriminatory. As another option to speed up these computations, max.cells.per.ident can be set. This will downsample each identity class to have no more cells than whatever this is set to. While there is generally going to be a loss in power, the speed increases can be significiant and the most highly differentially expressed features will likely still rise to the top.

This will allow you to see the gene expression for each cell 
```{r}
# find all markers of cluster 1
cluster1.markers <- FindMarkers(pbmc, ident.1 = 1, min.pct = 0.25)
head(cluster1.markers, n = 5)

# find all markers distinguishing cluster 5 from clusters 0 and 3
cluster5.markers <- FindMarkers(pbmc, ident.1 = 5, ident.2 = c(0, 3), min.pct = 0.25)
head(cluster5.markers, n = 5)

cluster2.markers <- FindMarkers(pbmc, ident.1 = 2, min.pct = 0.25)
head(cluster2.markers, n = 5)

cluster3.markers <- FindMarkers(pbmc, ident.1 = 3, min.pct = 0.25)
head(cluster3.markers, n = 5)

cluster4.markers <- FindMarkers(pbmc, ident.1 = 4, min.pct = 0.25)
head(cluster4.markers, n = 5)


```


Now that we have generated a UMAP and tSNE, we want to align the cell types
Using /Users/shailafye/Documents/summer2020/puram/cell_type
- this has the cell type for the 5902 cells
https://rpubs.com/mathetal/clusterlabel
https://satijalab.org/seurat/v3.1/pbmc3k_tutorial.html
Combine the cell type assignments w/ the cells 
- need the column name to be the cell type? 

This will allow you to see the gene expression for each cell 
```{r}

features <- c("MUSTN1")

RidgePlot(object = s_obj, features = features, ncol = 2)

VlnPlot(object = s_obj, features = features)

FeaturePlot(object = s_obj, features = features)

```

--------------------------------------------

ADD CELL TYPE ON TOP OF TSNE
- take coordinates and plot it with ggplot
- then plot w/ cell type

coordinates (https://github.com/satijalab/seurat/issues/1520)
- merge this with cell assign
- three columns
- x,y, and group (cell type)

then replot -- http://www.sthda.com/english/wiki/ggplot2-scatter-plots-quick-start-guide-r-software-and-data-visualization
x,y --> reg dot plot w/o color 
color = column name 


Now we take the coordinates of tsne and use ggplot to plot w/ cell type
- use our cell assign
- use published data cell type

Imposing cell type:
ggplot

- published cell type annotations
- our cell assign annotations for cell type

TSNE visualization

```{r TSNE ggplot, cell type}
#check if packages are installed
library(ggplot2)
install.packages("ggpubr")
library(ggpubr)
library(patchwork)
#file = "/Users/shailafye/Documents/summer2020/puram/tnse_attempt1.rds"
#extract coordinates of saved tsne plot above

s_obj <- readRDS(file = "/Users/shailafye/Documents/summer2020/tirosh/tirosh_tnse_attempt.rds")
coordinates = Embeddings(s_obj[["tsne"]])
coordinates=as.data.frame(coordinates)
head(coordinates)
coordinates$tSNE_1
coordinates[0] #cell

#dataframe with tsne coordinates, no cell type
df1 = data.frame(cell = coordinates[0], tsne1 = coordinates$tSNE_1, tsne2 = coordinates$tSNE_2)

#need to now use cell assign annotations
#add our generated cell type on top of tsne
cell_assign <-read.delim(file ="/Users/shailafye/Documents/summer2020/tirosh/cellAssign_tirosh_cell_type", stringsAsFactors=F)
class(cell_assign)
df2 = data.frame(cell = coordinates[0],
                 cell.type = cell_assign,
                       stringsAsFactors = F)
df1 #cells and tsne coordinates
df2 #cell type, cell_assign

#HN28_P15_D06_S330_comb	-26.03902952	-3.40291024	Vascular smooth muscle cells

data.plot = merge(df1, df2, by = 'row.names')
data.plot
names(data.plot)[4] <- "Cell"


#cell assign + cell type tsne plot
p1 = ggplot(data.plot, aes(x = tsne1, y = tsne2, cell=Cell)) +
  geom_point(aes(color = Cell, stroke = 0)) +
  theme_pubr(base_size = 10)+ ggtitle("Cell Assign Annotations (Tirosh)")

p1_colortest <- p1 + scale_color_manual(values = c("cyan1", "orchid", "darkorange", "mediumseagreen", "orange", "slateblue1","gold", "steelblue4", "violetred3"))+ ggtitle("Cell Assign Annotations (Tirosh)")  #10 -  "#CC79A7"

saveRDS(p1, file = "/Users/shailafye/Documents/summer2020/tirosh/tirosh_tsne_cellassign.rds")

#------------------
#published dataset/annotations
#"/Users/shailafye/Documents/summer2020/puram/published_celltype"
#published cell type --> edited so cancer cells are detected/recorded

published_annotations <- read.delim(file ="/Users/shailafye/Documents/summer2020/tirosh/melanoma_published_celltypes", stringsAsFactors=F)
head(published_annotations)

#published_annotations[published_annotations=="-Fibroblast"]<-"Fibroblast"
table(published_annotations)

df3 = data.frame(cell = coordinates[0],
                 cell.type = published_annotations,
                       stringsAsFactors = F)
df1 #cells and coordinates df
#combine cell type annotations with tsne coordinates
data.plot2 = merge(df1, df3, by = 'row.names')
data.plot2
names(data.plot2)[4] <- "Cell"
p3 = ggplot(data.plot2, aes(x = tsne1, y = tsne2, cell=Cell)) +
  geom_point(aes(color = Cell, stroke = 0)) +
  theme_pubr(base_size = 10)+ ggtitle("Published Cell Type Annotations (Tirosh)")

p3
p3_colortest <- p3 + scale_color_manual(values = c("cyan1", "lightpink1", "lightsalmon1", "darkorange", "slateblue1", "orange","mediumspringgreen", "olivedrab1", "gold", "steelblue4")) + ggtitle("Published Cell Type Annotations (Tirosh)")

saveRDS(p3, file = "/Users/shailafye/Documents/summer2020/tirosh/tirosh_tsne_published_celltype.rds")



```

-----------------------


UMAP VISUALIZATION

```{r UMAP}

pbmc <- readRDS(file = "/Users/shailafye/Documents/summer2020/tirosh/tirosh_umap_attempt.rds")
coordinates_umap = Embeddings(pbmc[["umap"]])
coordinates_umap=as.data.frame(coordinates_umap)
head(coordinates_umap)
coordinates_umap$tSNE_1
coordinates_umap[0] #cell

#dataframe with tsne coordinates, no cell type
df1_umap = data.frame(cell = coordinates_umap[0], UMAP_1 = coordinates_umap$UMAP_1, UMAP_2 = coordinates_umap$UMAP_2)

#need to now use cell assign annotations
#add our generated cell type on top of tsne
cell_assign <-read.delim(file ="/Users/shailafye/Documents/summer2020/tirosh/cellAssign_tirosh_cell_type", stringsAsFactors=F)
class(cell_assign)
df2_umap = data.frame(cell = coordinates_umap[0],
                 cell.type = cell_assign,
                       stringsAsFactors = F)
df1_umap #cells and tsne coordinates
df2_umap #cell type, cell_assign

#HN28_P15_D06_S330_comb	-26.03902952	-3.40291024	Vascular smooth muscle cells

data.plot_umap = merge(df1_umap, df2_umap, by = 'row.names')
data.plot_umap
names(data.plot_umap)[4] <- "Cell"

#cell assign + cell type tsne plot
p1_umap = ggplot(data.plot_umap, aes(x = UMAP_1, y = UMAP_2, cell=Cell)) +
  geom_point(aes(color = Cell, stroke = 0)) +
  theme_pubr(base_size = 10) + ggtitle("Cell Assign Annotations (Tirosh)")
p1_umap
p1_umap_colortest <- p1_umap + scale_color_manual(values = c("cyan1", "orchid", "darkorange", "mediumseagreen", "orange", "slateblue1","gold", "steelblue4", "violetred3"))+ ggtitle("Cell Assign Annotations (Tirosh)")  #10 -  "#CC79A7"


saveRDS(p1_umap, file = "/Users/shailafye/Documents/summer2020/tirosh/tirosh_umap_cellassign.rds")
#practice <- readRDS(file ="/Users/shailafye/Documents/summer2020/puram/umap_cellassign.rds")

#------------------
#published dataset/annotations
#"/Users/shailafye/Documents/summer2020/puram/published_celltype"
#published cell type --> edited so cancer cells are detected/recorded

published_annotations <- read.delim(file ="/Users/shailafye/Documents/summer2020/tirosh/melanoma_published_celltypes", stringsAsFactors=F)
head(published_annotations)

#published_annotations[published_annotations=="-Fibroblast"]<-"Fibroblast"
table(published_annotations)

df3_umap = data.frame(cell = coordinates_umap[0],
                 cell.type = published_annotations,
                       stringsAsFactors = F)
df1_umap #cells and coordinates df
#combine cell type annotations with tsne coordinates
data.plot2_umap = merge(df1_umap, df3_umap, by = 'row.names')
data.plot2_umap
names(data.plot2_umap)[4] <- "Cell"
p3_umap = ggplot(data.plot2_umap, aes(x = UMAP_1, y = UMAP_2, cell=Cell)) +
  geom_point(aes(color = Cell, stroke = 0)) +
  theme_pubr(base_size = 10)+ ggtitle("Published Cell Type Annotations  (Tirosh)")

p3_umap
p3_umap_colortest <- p3_umap + scale_color_manual(values = c("cyan1", "lightpink1", "lightsalmon1", "darkorange", "slateblue1", "orange","mediumspringgreen", "olivedrab1", "gold", "steelblue4")) + ggtitle("Published Cell Type Annotations  (Tirosh)")

#save this
saveRDS(p3_umap, file = "/Users/shailafye/Documents/summer2020/tirosh/tirosh_umap_published_celltype.rds")

```

